1-Draft/RFCnnnn-ForEach-Parallel-Cmdlet.md

```diff
@@ -0,0 +1,99 @@
+For the asynchronous case, the `ForEach-Object` cmdlet will immediately return a PowerShell job object that contains child jobs of each parallel execution.
+
+### Implementation details
+
+Implementation will be similar to the ThreadJob module.
+Script block execution will be run for each piped input on a separate thread and runspace.
+The number of threads that run at a time will be limited by a `-ThrottleLimit` parameter with a default value.
+Piped input that exceeds the allowed number of threads will be queued until a thread is available.
+For synchronous operation, a `-Timeout` parameter will be available that terminates the wait for completion after a specified time.
+Without a `-Timeout` parameter, the cmdlet will wait indefinitely for completion.
+
+### Synchronous parameter set
+
+Synchronous ForEach-Object -Parallel returns after all script blocks complete running or timeout
+
+```powershell
+ForEach-Object -Parallel -ThrottleLimit 10 -TimeoutSecs 1800 -ScriptBlock {} 
+```
+
+- `-Parallel` : parameter switch specifies fan-out parallel script block execution
+
+- `-ThrottleLimit` : parameter takes an integer value that determines the maximum number threads
+
+- `-TimeoutSecs` : parameter takes an integer that specifies the maximum time to wait for completion in seconds
+
+### Asynchronous parameter set
+
+Asynchronous ForEach-Object -Parallel immediately returns a job object for monitoring parallel script block execution
+
+```powershell
+ForEach-Object -Parallel -ThrottleLimit 5 -AsJob -ScriptBlock {}
+```
+
+- `-Parallel` : parameter switch specifies fan-out parallel script block execution
+
+- `-ThrottleLimit` : parameter takes an integer value that determines the maximum number threads
+
+- `-AsJob` : parameter switch returns a job object
+
+### Variable passing
+
+ForEach-Object -Parallel will support the PowerShell `$_` current piped item variable within each script block.
+It will also support the `$using:` directive for passing variables from script scope into the parallel executed script block scope.
+
+### Examples
+
+```powershell
+$computerNames = 'computer1','computer2','computer3','computer4','computer5'
+$logs = $computerNames | ForEach-Object -Parallel -ThrottleLimit 10 -TimeoutSecs 1800 -ScriptBlock {
+    Get-Logs -ComputerName $_
+}
+```
+
+```powershell
+$computerNames = 'computer1','computer2','computer3','computer4','computer5'
+$job = ForEach-Object -Parallel -ThrottleLimit 10 -InputObject $computerNames -AsJob -ScriptBlock {
+    Get-Logs -ComputerName $_
+}
+$logs = $job | Wait-Job | Receive-Job
+```
+
+```powershell
+$computerNames = 'computer1','computer2','computer3','computer4','computer5'
+$logNames = 'System','SQL'
+$logs = ForEach-Object -Parallel -InputObject $computerNames -ScriptBlock {
+    Get-Logs -ComputerName $_ -LogNames $using:logNames
+}
+```
+
+## Alternate Proposals and Considerations

```

Comment:
What if instead of looking at `ForEach-Object` for parallelism we scaled this out to allow parallelism to be used with any command?

For example, if we...
* ...had a new `CmdletBinding` attribute called `SupportsParallelism`.
* ...added `-Parallel`, `-ThrottleLimit`, `-TimeoutSecs`, and `-AsJob` common parameters to any advanced function or cmdlet that has `SupportsParallelism` set to `$true` in the `CmdletBinding` attribute (supporting the different common parameter sets that are proposed above, such that `-AsJob` could not be used with `-TimeoutSecs`).
* ...updated the `$using:` variable prefix so that it can be used in any _process_ block inside of a command that has `SupportsParallelism` set to `$true` (note: this would also need to allow `$using:` to work when the command is invoked without parallelism, so that the functionality would only need to be written once; the `$using:` prefix would simply identify a variable that is pulled into the parallel runspace/job when the command is invoked in parallel.

Then we could...
* ...add the `SupportsParallelism` attribute to `ForEach-Object` invocations to get the functionality we're talking about here.
* ...add the `SupportsParallelism` attribute to any other command that is written to support parallelism.
* ...simplify command invocation when you want parallelism a great deal, which would keep `ForEach-Object` use to where it is really needed.
* ...maintain support for `begin` and `end` blocks (which are still useful -- even with parallelism, some resources such as connections, credentials, or collections to accumulate results may be needed and shared among the different threads/jobs).
* ...discover commands that support parallelism (and therefore those that do not).
* ...make multitasking much more of a "thing" in PowerShell.

Taking the examples provided above, they could be simplified as follows:

```powershell
$computerNames = 'computer1','computer2','computer3','computer4','computer5'
$logNames = 'System','SQL'

# Get logs from computers in parallel, throttled with a timeout
$logs = Get-Logs -ComputerName $computerNames -Parallel -ThrottleLimit 10 -TimeoutSecs 1800

# Get logs from computers in parallel using background jobs
$logs = Get-Logs -ComputerName $computerNames -Parallel -ThrottleLimit 10 -AsJob | Receive-Job -Wait

# Get System and SQL logs from computers in parallel with no throttle or timeout
$logs = Get-Logs -ComputerName $computerNames -LogNames $logNames -Parallel
```

For folks who are going to look at this and complain about the lack of downlevel support in this functionality, I'd like to draw your attention to [this RFC](https://github.com/PowerShell/PowerShell-RFC/pull/202). That won't help projects that require a downlevel version, but PowerShell needs to start offering a lot more motivation for users to upgrade to newer releases, and this would be some pretty great motivation.

Reply:
An interesting idea, and would be a lot of work to cover all built-in cmdlets.  A variation is to build on @BrucePay  work with the `&` operator to force execution as background job.

```powershell
Get-Process -id $pid &
Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
--     ----            -------------   -----         -----------     --------             -------
5      Job5            BackgroundJob   Running       True            localhost            Microsoft.PowerShell.Man…
```
Maybe this can be extended to optionally fan-out using ThreadJobs.

This is outside the scope of this RFC, but is something we could look into in the future.

Reply:
It shouldn't really be a lot of work to cover all built-in cmdlets. For the first cut you could just lay the foundation and update `ForEach-Object` to use `SupportsParallelism` (or just call that `SupportsParallel` for short and to match the `-Parallel` common parameter name). The other built-in cmdlets could be updated over time. That keeps focus on what this RFC is trying to do (add parallelism to ForEach-Object) while doing so in such a way that it can scale to any cmdlet/advanced function.

Reply:
Also regarding the `&` background operator and thread jobs, see [PowerShell Issue #9873](https://github.com/PowerShell/PowerShell/issues/9873).

Reply:
Are you thinking of this in the same simple sense as `SupportsShouldProcess` where all you're doing is opting in to a few common parameters and magic variables which you may --or may not, actually-- use (properly)?

Reply:
Yes. I think the complexity of using variables with parallelism needs to be dealt with no matter what, but my hope is that the complexity can be reduced for parameters, and maybe that certain collection enclosures (see RFC #204) can be added to PowerShell to make working with collections across thread boundaries easier to do properly.

Reply:
Good ideas.  Especially making $using: available everywhere. 

You would then have foreach-object as the first worked example of using the new directive, and how it should be done. It's then for others to decide if (e.g) where-object, or Start-Process would benefit from it or not. 

Then you provide a template for how to setup a thread pool, add threads and collect results ? I'm just weighing up whether having a "supportsParallelism" attribute would make any difference to whether I added parallelism to a function - I've referred to Start-Parallel, but I've never felt the need to bring that functionality into other functions I've written.

Reply:
While I was running errands this evening, another thought occurred to me. Since PowerShell commands have different parameter sets, some of which are quite different from one another and used for different purposes, it would probably be useful to be able to identify which parameter sets support parallelism. i.e Instead of checking for the presence of `SupportsParallelism` as a boolean value, we could check if it is present without assignment (just like we do for `SupportsShouldProcess`), in which case it would apply to all parameter sets, or command authors could assign a collection of parameter sets to it to indicate which parameter sets support parallelism (e.g. `[CmdletBinding(DefaultParameterSetName='Default',SupportsParallelism='PSet1','PSet2')]`).

Reply:
@KirkMunro I suggest to split the (large) discussion and to pull new RFC with your great proposal for general parallelism in PowerShell.

Reply:
Yeah, this probably is worth splitting out. There's no reason why we can't abstract the set of parameters which belong in this new common parameter extension set from whatever ends up being implemented when this RFC is done, and then retrofit the attribute onto the command, right?

----

What if we could assert that this command would use the fastest, most consistent, most light-weight implementation of parallel threads possible -- would we still want other people to attempt their own implementations?  That is, do we really want parallel processing built-in to many different commands, in many different ways, if we can _simply_ wrap any command in `ForEach-Parallel` -- and even though their implementations may not be as good?

Bear in mind that a lot of commands _already_ support parallelization when used with a `-ComputerName [string[]]` parameter, but do so by fan-out remoting ...

I would almost rather an automatic implementation: `[ParallelCmdletBinding()]` on a script command means your `process` block will automatically be parallelized using thread jobs, and all `local` variables will be copied over and back out to the end block ...

Even without that, _I still like the idea of a set of parameters for parallelization_, but I suspect the only implementations will be the ones that are already out there (i.e. Start-Parallel, Invoke-Parallel, and other variations on the theme) with each one providing a slightly different implementation of parallelization, and possibly needing to tweak the parameters because of the implementations.

In fact, the commonality might be misleading if every command has it's own implementation of parallelization. If all parallel commands have a `-ThrottleLimit` and people learn to assume that this parameter means "number of threads" ... what happens when a command is using processes, or worse, what if they provides parallelization based on fan-out remoting? In that case, will it just choose to treat `-ThrottleLimit` as a number of instances per-computer?

P.S. For what it's worth, it's clear from @PaulHigin's comments about the begin/end parameters on ForEach-Object that we _would_ need to specify which parametersets support parallelization. I don't think anything like that has been implemented in the other common parameter extensions, so this might end up being relatively complicated.

Reply:
Agree !

Reply:
I've already started a separate RFC for this to see where that leads. Thanks for sharing these thoughts @Jaykul, I had some of them already, but not quite all. Automatic implementation is the goal, for consistency and performance, but I'm still thinking it would be bound to specific named parameter sets, or all parameter sets if you don't name them. Part of what I hope happens here is to "corral the troops" so to speak, and bring a consistent implementation to the table that commands can adopt.

Reply:
What about adding a new, mutually exclusive [to process] processasync/parallelprocess block in cmdlets/functions? 

```
function foo {
    begin { }
    processasync { }
    end { }
}
```
It's so easy to write that out.

Reply:
@oising The last thing you suggested (CmdletBinding configuration) is exactly what started this part of the discussion, albeit with different terminology (`[CmdletBinding(SupportsParallel=...)]`). I'm also considering `processasync` and the corresponding `ProcessRecordAsync` while I work through the complementary RFC.

Reply:
Cool. I'm leaning towards the idea of just leaving the process keyword alone (e.g. no new keywords like processasync), but in the .NET API, surface ProcessRecordAsync, and have them materialize as threadjobs in the console.

Reply:
Here's the complementary RFC I wrote up today: https://github.com/PowerShell/PowerShell-RFC/pull/206.